crates/subspace-proof-of-time/src/pot_aes.rs

```diff
@@ -0,0 +1,238 @@
+use alloc::vec::Vec;
+#[cfg(any(feature = "parallel", test))]
+use rayon::prelude::*;
+use subspace_core_primitives::{PotBytes, PotCheckpoint, PotKey, PotSeed};
+
+/// Creates the AES based proof.
+pub(crate) fn create(
+    seed: &PotSeed,
+    key: &PotKey,
+    num_checkpoints: u8,
+    checkpoint_iterations: u32,
+) -> Vec<PotCheckpoint> {
+    let key = GenericArray::from(PotBytes::from(*key));
+    let cipher = Aes128::new(&key);
+    let mut cur_block = GenericArray::from(PotBytes::from(*seed));
+
+    let mut checkpoints = Vec::with_capacity(num_checkpoints as usize);
+    for _ in 0..num_checkpoints {
+        for _ in 0..checkpoint_iterations {
+            // Encrypt in place to produce the next block.
+            cipher.encrypt_block(&mut cur_block);
+        }
+        let bytes: PotBytes = cur_block.into();
+        checkpoints.push(PotCheckpoint::from(bytes));
+    }
+    checkpoints
+}
+
+/// Verifies the AES based proof sequentially.
+#[cfg(any(not(feature = "parallel"), test))]
+pub(crate) fn verify_sequential(
+    seed: &PotSeed,
+    key: &PotKey,
+    checkpoints: &[PotCheckpoint],
+    checkpoint_iterations: u32,
+) -> bool {
+    let key = GenericArray::from(PotBytes::from(*key));
+    let cipher = Aes128::new(&key);
+    let mut cur_block = GenericArray::from(PotBytes::from(*seed));
+
+    for checkpoint in checkpoints {
+        // Encrypt in place checkpoint_iterations times
+        // and compare with the checkpoint.
+        for _ in 0..checkpoint_iterations {
+            cipher.encrypt_block(&mut cur_block);
+        }
+
+        let checkpoint_block = GenericArray::from(PotBytes::from(*checkpoint));
+        if cur_block != checkpoint_block {
+            return false;
+        }
+    }
+    true
+}
+
+/// Verifies the AES based proof in parallel.
+#[cfg(any(feature = "parallel", test))]
+pub(crate) fn verify_parallel(
+    seed: &PotSeed,
+    key: &PotKey,
+    checkpoints: &[PotCheckpoint],
+    checkpoint_iterations: u32,
+) -> bool {
+    let key = GenericArray::from(PotBytes::from(*key));
+    let cipher = Aes128::new(&key);
+
+    // Create the cipher pairs to be evaluated
+    let mut pairs = Vec::new();
+    let mut cur_block = GenericArray::from(PotBytes::from(*seed));
+    for checkpoint in checkpoints {
+        let checkpoint_block = GenericArray::from(PotBytes::from(*checkpoint));
+        pairs.push((cur_block, checkpoint_block));
+        cur_block = checkpoint_block;
+    }
+
+    // Evaluate the pairs in parallel.
+    let results: Vec<bool> = pairs
+        .par_iter_mut()
+        .map(|(input, expected)| {
+            let mut input = *input;
+            // Encrypt in place checkpoint_iterations times
+            // and compare with the expected output.
+            for _ in 0..checkpoint_iterations {
+                cipher.encrypt_block(&mut input);
+            }
+            input == *expected
+        })
+        .collect();
+
+    results.iter().all(|result| *result)
```

Comment:
Feel free to consider this as a future optimization, but this is inefficient in a number of ways:
* vector does not have preallocated size
* pairs do not need to be stored in a vector (use [par_windows](https://docs.rs/rayon/latest/rayon/slice/trait.ParallelSlice.html#method.par_windows) instead)
* results do not need to be collected into a vector and iterated again either, it is sufficient to find the first mismatch and abort computation

So this should be `par_windows` with not explicit allocations. Can be refactored later though, just sharing the way I think about the code when I read it.

Reply:
This is less than 256 entries now, not so much worried about it, but will certainly revisit after we have some basic stuff working